common/src/main/java/io/druid/collections/StupidPool.java

```diff
@@ -74,25 +72,24 @@ public synchronized T get()
     }
 
     @Override
-    public synchronized void close() throws IOException
+    public void close() throws IOException
     {
-      if (closed) {
+      if (!closed.compareAndSet(false, true)) {
         log.warn(new ISE("Already Closed!"), "Already closed");
         return;
       }
-
-      synchronized (objects) {
-        closed = true;
-        objects.addLast(object);
+      if (!objects.offer(object)) {
+        log.warn(new ISE("Queue offer failed"), "Could not offer object [%s] back into the queue", object);
       }
     }
 
     @Override
     protected void finalize() throws Throwable
     {
-      if (! closed) {
+      if (!closed.get()) {
         log.warn("Not closed!  Object was[%s]. Allowing gc to prevent leak.", object);
       }
+      super.finalize();
```

Comment:
if we need to add it, it should probably be in a finalize block to ensure it gets run no matter what happens